src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,297 @@
+            public ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode literalExpressionSyntaxNode)> ParametersWithLiteralArgument { get; }
+
+            private MethodParametersInfo(
+                ImmutableArray<(IParameterSymbol, SyntaxNode)> parametersWithIdentifierArgument,
+                ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode variableDeclarationNode)> parametersWithVariableDeclarationArgument,
+                ImmutableArray<(IParameterSymbol, ImmutableArray<SyntaxNode>)> parametersNeedGenerateDeclarations,
+                ImmutableArray<IParameterSymbol> parametersWithDefaultValue,
+                ImmutableArray<(IParameterSymbol parameterSymbol, SyntaxNode literalExpressionSyntaxNode)> parametersWithLiteralArgument)
+            {
+                ParametersWithIdentifierArgument = parametersWithIdentifierArgument;
+                ParametersWithVariableDeclarationArgument = parametersWithVariableDeclarationArgument;
+                ParametersNeedGenerateDeclarations = parametersNeedGenerateDeclarations;
+                ParametersWithDefaultValue = parametersWithDefaultValue;
+                ParametersWithLiteralArgument = parametersWithLiteralArgument;
+            }
+
+            public static MethodParametersInfo GetMethodParametersInfo(
+                AbstractInlineMethodRefactoringProvider<TInvocationSyntaxNode, TExpressionSyntax, TArgumentSyntax> inlineMethodRefactoringProvider,
+                ISyntaxFacts syntaxFacts,
+                SemanticModel semanticModel,
+                SyntaxNode calleeInvocationSyntaxNode,
+                IMethodSymbol calleeMethodSymbol,
+                CancellationToken cancellationToken)
+            {
+                // Classify the parameters into:
+                // 1. If the parameter accept an identifier from Caller, then use this identifier to replace all the occurence of the parameter.
+                // (So that after inlining, the same identifier could be found in the caller)
+                // 2. If the parameter accept literal/no argument but has default value, replace all the occurence of the parameter by using the
+                // literal expression
+                // 3. If the argument has variable declarations ('out var' in C#), then insert an additional declaration. Also replace all the
+                // occurence of parameter by using this identifier.
+                // 4. For the rest of the parameters (might include method invocation and different expressions), use the parameter's name to
+                // generate a declaration for them and insert that into caller.
+                var allArguments = syntaxFacts.GetArgumentsOfInvocationExpression(calleeInvocationSyntaxNode);
+                var parameterSymbolAndArguments = allArguments
+                    .Select(argument => (
+                        parameterSymbol: inlineMethodRefactoringProvider.GetParameterSymbol(semanticModel, (TArgumentSyntax)argument,
```

Comment:
> would it be easier/more-conveinent to use IOperation?
Port from your one original comment because I have refactored the code to a different location. @CyrusNajmabadi 

I don't have a strong preference to use this method call or IOperation. I thought we have a talk about this and you mentioned that IOperation only works fine when everything is correct. So I use this here in to be more fault-tolerant.  But I am not sure if this really make sense